Context completion

jneem opened the pull request:
This adds completions from things that are "in scope," liberally interpreted. This includes things that are actually in scope, but also things that we can tell will be "brought into" scope by merges and contract annotations.

Fixes #1499. This also makes (at least according to the current regression tests) the new completer a superset of the old one. I'll be planning to remove the old one soonish. (After 1.2, obviously, and also after adding some more comprehensive tests.)

One behavioral question I have is whether it makes sense to include this scope-based completion if we detect that they are in the middle of a record path. The old completer did, so I am doing it also. But since we offer more completions than the old completer, maybe it makes too many false positives?

vkleen reviewed (approved):
This still needs to be rebased, but other than that LGTM.

yannham commented on lsp/nls/tests/snapshots/main__lsp__nls__tests__inputs__completion-dict.ncl.snap:
```diff
@@ -2,6 +2,6 @@
 ---
-[foo]
-[foo]
+[foo, x]
+[Dict, foo, x]
```
Same here, why we have completion for both variable from the environment and dictionary fields? It seems like we are conflating completion from the current environment (variables in scope) and field completion (triggered by a `.`). But those are arguably different, and might be a regression: when I write `myrecord.`, variables in scope have no reason to be valid input here - we should only suggest completion from what fields can be inferred from `myrecord`. Same for context completion. On the other hand, non-record completion should only show things in scope.

jneem replied:
I think I agree with you (it's what I intended to bring up in the last paragraph of the PR description). The previous nls behavior was using both completion sources when completing a record path, so I was just copying that behavior.

I'll make the change to turn of the variables-in-scope completion when completing record fields, and we can see if we like the results.

yannham replied:
Ah, I see, I think I didn't really understand that paragraph, and once I had sufficient context could I had just forget about it :sweat_smile:

jneem commented on lsp/nls/src/linearization/mod.rs:
```diff
@@ -30,6 +30,57 @@ pub mod interface;
+        let mut traverse_merge =
+            |rt: &RichTerm, parent: &Option<RichTerm>| -> TraverseControl<Option<RichTerm>, ()> {
+                if let Some(parent) = parent {
+                    table.insert(RichTermPtr(rt.clone()), parent.clone());
```
Good point, we are not dropping the table yet. I'll fix that.

yannham replied:
Another solution is to use weak pointers

jneem commented on lsp/nls/src/linearization/mod.rs:
```diff
@@ -30,6 +30,57 @@ pub mod interface;
 
+#[derive(Clone, Debug)]
+pub struct ParentLookup {
+    table: HashMap<RichTermPtr, RichTerm>,
```
I think we don't care about the position, and `cargo test` agrees so I'll drop it. I'll do the pointer-chasing in a follow-up, if that's ok. I'd like to prioritize getting rid of the linearizer.

yannham replied:
Sounds like a plan :+1:

jneem commented on lsp/nls/tests/snapshots/main__lsp__nls__tests__inputs__completion-basic.ncl.snap:
```diff
@@ -3,16 +3,16 @@ source: lsp/nls/tests/main.rs
-[foo, really, verified, version]
-["has a space", lalala]
-[falala]
+[a, b, config, foo, verified, version]
```
LSP lines (and characters) start from zero, so this is pointing at the "a = config.version" line. Character 12 points at "config". At that point, `a`, `b`, and `config` are in scope, and  I think this is what the snapshot shows (it's the line above the one you commented on). The next query is for character 15, which is after "config.", so it correctly has `config`'s fields in scope also.

yannham replied:
> LSP lines (and characters) start from zero

For some reason I was hard convinced that they were starting at 1, my bad.

jneem commented:
Ok, I rebased and I also changed the behavior to use either context completion or env completion, but not both at once. At the same time, I disabled the old completer by default so that the new behavior is easier to see. What do you think of the test output now, @yannham ?

yannham reviewed (approved):
Looks good when only selecting one source of completion :+1: 

I wonder if we have any tests for context completion (as in, completion when defining a field inside a record literal based on parents)? I only grepped quickly over the completion inputs for test but only saw dot-triggered completion mostly. Doesn't have to be in this PR, but this could be interesting to monitor as well.

jneem commented:
Good point, there were one or two cases in the old tests that hit env completion, but I added a few more and already found (and fixed) an issue :sweat_smile:
